### ISSUE TITLE
[restrict-template-expressions] does not correctly resolve string properties

### ISSUE THREAD

**Issue by cmcnicholas** [typescript-eslint/typescript-eslint#4070]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": "error"
  }
}
```

```TS
const dragon = {
  noise: 'roar!',
};

console.log(`dragon goes ${dragon.noise}`);
```

**Expected Result**

The above should not be flagged as a linting error.

**Actual Result**

It flags the above property as a linting error, changing the code to the following makes it work but is convoluted:

```TS
const dragon = {
  noise: 'roar!',
};

const noise: string = dragon.noise;
console.log(`dragon goes ${noise}`);

// or even
console.log(`dragon goes ${String(dragon.roar)}`);
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.33.0` |
| `@typescript-eslint/parser`        | `4.33.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `16.6.1` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.
Please provide a repro repo.

**Comment by cmcnicholas** [comment]
I think I have managed to narrow this down, when working in a typescript project with project references do the referenced projects require to be built for the plugin to infer their type?

https://github.com/cmcnicholas/typescript-eslint-restrict-template-expressions-project-references

I have setup a basic repo with 2 projects, `project-b` references `project-a`.

Steps to repro:

1. checkout repo
2. `cd ./projects/project-b`
3. `npm run lint` - see this fails in `./projects/project-b/src/main.ts`
4. `npm run build` - builds project b and referenced project a
5. `npm run lint` - see this passes

To restart the test, delete the `dist` dirs created, restarting VSCode between operations will result in pass/fail from the IDE for linting too.

**Comment by JoshuaKGoldberg** [comment]
Indeed:

```
  10:30  error  Invalid type "any" of template literal expression  @typescript-eslint/restrict-template-expressions
```

Your issue is that you need to build all project references (aka dependencies) of a composite project before using it. If you try to run `tsc` in `./projects/project-b` before any `npm run build`:

```
C:\Code\typescript-eslint-restrict-template-expressions-project-references\projects\project-b>tsc 
error TS6305: Output file 'C:/Code/typescript-eslint-restrict-template-expressions-project-references/projects/project-a/dist/src/Dragon.d.ts' has not been built from source file 'C:/Code/typescript-eslint-restrict-template-expressions-project-references/projects/project-a/src/Dragon.ts'.
  The file is in the program because:
    Matched by include pattern '../project-a/src/**/*.ts' in 'C:/Code/typescript-eslint-restrict-template-expressions-project-references/projects/project-b/tsconfig.json'

  tsconfig.json:20:30
    20   "include": ["src/**/*.ts", "../project-a/src/**/*.ts"],
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~
    File is matched by include pattern specified here.


Found 1 error.
```

This is not an issue with typescript-eslint.

**Suggestion**: use a root-level `tsconfig.json` that has `files: []` and a `references` list to coordinate building packages. Example: https://github.com/Codecademy/use-selector-with/blob/2c0bcf7/tsconfig.json

**Comment by cmcnicholas** [comment]
this makes sense thanks, the example is quite diluted from what I really have but I guess I will need to figure out how to build a Vue project that is being used as a project reference, it isn't directly buildable with `tsc` which is where this issue is arising from within a shared project that contains reusable Vue components alongside TS.

Thanks again.
